File: LeetCode_Algorithm/1200/1200.py
from typing import List

class Solution:
    def minimumAbsDifference(self, arr: List[int]) -> List[List[int]]:
        # sol1 O(nlogn) O(n) 415~492ms 69~47%
        # 소팅 후 각 차이를 계산
        # 만약 차이가 더 적다면 초기화 후 계산
        # arr.sort()
        # ans, diff = [], 100090000000
        # for idx in range(len(arr)-1):
        #     if diff > arr[idx+1]-arr[idx]:
        #         diff = arr[idx+1]-arr[idx]
        #         ans = [[arr[idx], arr[idx+1]]]
        #     elif diff == arr[idx+1]-arr[idx]:
        #         ans.append([arr[idx], arr[idx+1]])
        # return ans

        # sol2  O(m+n), O(m) 685~717ms 11~8%
        # counting sort
        # 음수도 있으므로 가장 작은 수를 더해서 배열을 완성시킨다.
        min_num = min(arr)
        for idx in range(len(arr)): 
            arr[idx] += abs(min_num)

        # 카운팅 배열 생성
        counts = [0] * (max(arr)+1)
        for idx in arr:
            counts[idx] += 1


        # 카운트 배열에서 두 개의 포인터를 만들어 간격을 잰다.
        # 최소의 간격일 경우 배열을 초기화하고 다시 계산한다.
        ans = []
        idx_i, idx_j = -1, -1
        diff = 10000000000000
        for idx in range(len(counts)):
            if counts[idx] >= 1:
                idx_i, idx_j = idx_j, idx
                if idx_i >= 0 and idx_j >= 0:
                    num1 = idx_i-abs(min_num)
                    num2 = idx_j-abs(min_num)

                    if diff > num2-num1:
                        ans = [[num1, num2]]
                        diff = num2-num1

                    elif diff == num2-num1:
                        ans.append([num1, num2])
        return ans

 


print(Solution().minimumAbsDifference([3,8,-10,23,19,-4,-14,27]))

File: LeetCode_Algorithm/1200/1200.js
var minimumAbsDifference = function(arr) {
  // sol1 232~253ms 45~32%
  
  // arr.sort(function(a,b){return a-b})
  // var diff = 100000000000
  // var ans = []
  
  // for (let idx=1; idx<arr.length; idx++) {
  //   if ((arr[idx]-arr[idx-1]) < diff) {    
  //     ans = [[arr[idx-1], arr[idx]]]
  //     diff = arr[idx]-arr[idx-1]
  //   }
  //   else if ((arr[idx]-arr[idx-1]) == diff) {
  //     ans.push( [arr[idx-1], arr[idx]] )
  //   }
  // }
  // return ans

  // sol2 186~232ms 78~45%
  var minNum = Math.min(...arr)
  if (minNum < 0) {minNum *= -1}
  
  for (let idx=0; idx<arr.length; idx++){
    arr[idx] += minNum
  }

  var counts = new Array(Math.max(...arr)+1).fill(0)
  for (const num of arr){
    counts[num] = 1
  }

  var ans = []
  var idxI = -1; var idxII = -1
  var diff = 100000000000000
  for(let idx=0; idx<counts.length; idx++){
    if (counts[idx] == 1){
      idxI = idxII; idxII = idx
      if (idxI >= 0 && idxII >= 0) {
        var numI = idxI-minNum; var numII = idxII-minNum;
        if (numII-numI < diff){
          diff = numII-numI
          ans = [[numI, numII]]
        }
        else if (numII-numI == diff) {
          ans.push([numI, numII])
        }
      }
    }
  }
  return ans
  
  
};
console.log(minimumAbsDifference([3,8,-10,23,19,-4,-14,27]))